tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/service/impl/TrainingNumberServiceImpl.java

```diff
@@ -117,4 +136,293 @@ public void delete(Long id) {
+    log.info("Populating training number for programme membership '{}'.",
+        programmeMembership.getUuid());
+
+    if (isExcluded(programmeMembership)) {
+      return;
+    }
+
+    String parentOrganization = getParentOrganization(programmeMembership);
+    String specialtyConcat = getSpecialtyConcat(programmeMembership);
+    String referenceNumber = getReferenceNumber(programmeMembership.getPerson());
+    String suffix = getSuffix(programmeMembership);
+
+    TrainingNumber trainingNumber = new TrainingNumber();
+    trainingNumber.setTrainingNumber(
+        parentOrganization + "/" + specialtyConcat + "/" + referenceNumber + "/" + suffix);
+    programmeMembership.setTrainingNumber(trainingNumber);
+    log.info("Populated training number: {}.", trainingNumber);
+  }
+
+
+  /**
+   * Get the parent organization for the given programme membership.
+   *
+   * @param programmeMembership The programme membership to calculate the parent organization for.
+   * @return The calculated parent organization.
+   */
+  private String getParentOrganization(ProgrammeMembership programmeMembership) {
+    String managingDeanery = programmeMembership.getProgramme().getOwner();
+    log.info("Calculating parent organization for managing deanery '{}'.", managingDeanery);
+    String parentOrganization = null;
+
+    if (managingDeanery != null) {
+      switch (managingDeanery) {
+        case "Defence Postgraduate Medical Deanery":
+          parentOrganization = "TSD";
+          break;
+        case "Health Education England East Midlands":
+          parentOrganization = "EMD";
+          break;
+        case "Health Education England East of England":
+          parentOrganization = "EAN";
+          break;
+        case "Health Education England Kent, Surrey and Sussex":
+          parentOrganization = "KSS";
+          break;
+        case "Health Education England North Central and East London":
+        case "Health Education England South London":
+        case "Health Education England North West London":
+        case "London LETBs":
+          parentOrganization = "LDN";
+          break;
+        case "Health Education England North East":
+          parentOrganization = "NTH";
+          break;
+        case "Health Education England North West":
+          parentOrganization = "NWE";
+          break;
+        case "Health Education England South West":
+          parentOrganization = "SWN";
+          break;
+        case "Health Education England Thames Valley":
+          parentOrganization = "OXF";
+          break;
+        case "Health Education England Wessex":
+          parentOrganization = "WES";
+          break;
+        case "Health Education England West Midlands":
+          parentOrganization = "WMD";
+          break;
+        case "Health Education England Yorkshire and the Humber":
+          parentOrganization = "YHD";
+          break;
+        default:
+          break;
+      }
+    }
```

Comment:
This is ugly and not ideal, especially with looming name changes. However there is currently no mapping available and until recently it wasn't possible to directly map without additional logic.
Now that SW have standardised on a single prefix code it seems sensible to record the prefixes in the TIS reference data and perform a lookup here instead of hardcoding.

I'll write a ticket for that to be done separate, with a sub-task to clean this code up!

Reply:
May be something like this?
```
private static final Map<String, List<String>> heeDbToOwnerMap = ImmutableMap.<String, List<String>>builder()
      //London LETBs match any of the london DBC's
      .put("1-1RUZV1D",
          Lists.newArrayList("Health Education England Kent, Surrey and Sussex", "London LETBs"))
      .put("1-1RUZV6H",
          Lists.newArrayList("Health Education England North West London", "London LETBs"))
      .put("1-1RUZV4H", Lists
          .newArrayList("Health Education England North Central and East London", "London LETBs"))
      .put("1-1RSSQ5L", Lists.newArrayList("Health Education England South London", "London LETBs"))

      .put("1-1RSSPZ7", Lists.newArrayList("Health Education England East Midlands"))
      .put("1-1RSSQ05", Lists.newArrayList("Health Education England East of England"))
      .put("1-1RSSQ1B", Lists.newArrayList("Health Education England North East"))
      .put("1-1RSSQ6R", Lists.newArrayList("Health Education England Thames Valley"))
      .put("1-1RSG4X0", Lists.newArrayList("Health Education England Yorkshire and the Humber"))
      .put("1-1RUZUYF", Lists.newArrayList("Health Education England West Midlands"))
      .put("1-1RUZUVB", Lists.newArrayList("Health Education England South West"))
      .put("1-1RUZUSF", Lists.newArrayList("Health Education England Wessex"))
      .put("1-1RSSQ2H", Lists.newArrayList("Health Education England North West"))
      .put("1-25U-830", Lists.newArrayList("Northern Ireland Medical and Dental Training Agency"))
      .build();
```

Reply:
We could just bump the Java version and I can make it pretty 😉 

```java
String parentOrganization = managingDeanery == null ? null : switch (managingDeanery) {
  case "Defence Postgraduate Medical Deanery" -> "TSD";
  case "Health Education England East Midlands" -> "EMD";
  case "Health Education England East of England" -> "EAN";
  case "Health Education England Kent, Surrey and Sussex" -> "KSS";
  case "Health Education England North Central and East London",
      "Health Education England South London",
      "Health Education England North West London",
      "London LETBs" -> "LDN";
  case "Health Education England North East" -> "NTH";
  case "Health Education England North West" -> "NWE";
  case "Health Education England South West" -> "SWN";
  case "Health Education England Thames Valley" -> "OXF";
  case "Health Education England Wessex" -> "WES";
  case "Health Education England West Midlands" -> "WMD";
  case "Health Education England Yorkshire and the Humber" -> "YHD";
  default -> null;
};
```